content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+of species and each tree will be different. This collective group will work
+together to stay healthy. As a whole, a given forest will have a sort of
+homogeneity, a character. Now, burn everything down and use a single seed to
+rebuild your forest. This results in a monoculture, which is poor and less
+robust to the environment.
+
+Setting the seed of a random number generator is just like that. There is
+nothing random anymore, you are breaking your random number generator, breaking
+your forest. You see often in **code** the exact same seeds: 0, 123456.
+In fact, if you do this, you are using a fixed sequence of points which
+effectively has no guarantee over the quality of the points.
+
+Consider a few statistical distributions to illustrate this.
+
+_Note that I used 1024 points which is fairly reasonable for a 1 dimensional
+problem._
+
+```python
+from collections import namedtuple
+
+seed = 123456  # Please do NOT use a seed: this is to illustrate a wrong usage
+rng = np.random.default_rng(seed)
+
+# samples
+ns = 1_024
+
+dist_ = namedtuple("dist", ["name", "sample", "pdf"])
+dists = [
+    dist_("uniform", rng.random(ns), stats.uniform),
+    dist_("normal", rng.standard_normal(ns), stats.norm),
+    dist_("gamma", rng.gamma(2., size=ns), stats.gamma(2.))
+]
+
+# visualization
+fig, axs = plt.subplots(1, 3, figsize=(12, 4))
+
+for i, dist in enumerate(dists):
+    _, bins, _ = axs[i].hist(dist.sample, bins=64, density=True)
+    x = np.linspace(min(bins), max(bins), 100)
+    axs[i].plot(x, dist.pdf.pdf(x), 'r-', lw=5, alpha=0.6, label=dist.name)
+    axs[i].legend()
+```
+
+![
+Histograms for uniform, normal and gamma distribution using 1024 samples.
+They are compared with the underlying distributions and a poor match. 
+](distributions_seeded.png)
+
+Does that look good to you? It is not. If you need to have a determined
+sequence of points, there are other technics that would generate almost
+perfect histograms only using a fraction of the number of points which is used
+here. See [this article]({{< relref "../../scipy/qmc-basics" >}})
+for more information on this.
+
+But why does it matter? How bad this is will depend on your application.
+If you don't, statistical properties are lost. This is hard to get because for
+most people, random is random. So taking an arbitrary draw of sample should not
+matter. And it does not **if** you replicate your analysis using different
+**random** seeds. This is the part everyone is forgetting. Whether you are
+optimizing something or calculating some statistics, you could get (un)lucky
+if you do this just once.
+
+If your application needs random numbers, then do use random numbers and play
+with the fact that they are random. You do science or anything critical
+using these randomly generated numbers, then please stay away from seeds.
+
+Some will argue that seeding is just used by developers for testing, well it is
```

Comment:
I think this could be rewritten:

"Although it can be argued that seeding is just used by developers for testing, this is not always the case."

Reply:
The attempts during this blog to use a more casual language may end up adding unnecessary information that just makes the article longer. I believe that the lines 232-236 can be summarized in a single sentence, for instance.